dpctl/tensor/libtensor/source/reductions.cpp

```diff
@@ -0,0 +1,469 @@
+    }
+
+    if (supports_atomics && (reduction_nd == 1) &&
+        (simplified_reduction_src_strides[0] == 1) && (iteration_nd == 1) &&
+        ((simplified_iteration_shape[0] == 1) ||
+         ((simplified_iteration_dst_strides[0] == 1) &&
+          (static_cast<size_t>(simplified_iteration_src_strides[0]) ==
+           reduction_nelems))))
+    {
+        auto fn =
+            sum_over_axis_contig_atomic_dispatch_table[src_typeid][dst_typeid];
+        if (fn != nullptr) {
+            size_t iter_nelems = dst_nelems;
+
+            sycl::event sum_over_axis_contig_ev =
+                fn(exec_q, iter_nelems, reduction_nelems, src.get_data(),
+                   dst.get_data(), iteration_src_offset, iteration_dst_offset,
+                   reduction_src_offset, depends);
+
+            sycl::event keep_args_event = dpctl::utils::keep_args_alive(
+                exec_q, {src, dst}, {sum_over_axis_contig_ev});
+
+            return std::make_pair(keep_args_event, sum_over_axis_contig_ev);
+        }
+    }
+
+    using dpctl::tensor::kernels::sum_reduction_strided_impl_fn_ptr;
+    sum_reduction_strided_impl_fn_ptr fn = nullptr;
+
+    if (supports_atomics) {
+        fn =
+            sum_over_axis_strided_atomic_dispatch_table[src_typeid][dst_typeid];
+    }
+
+    if (fn == nullptr) {
+        // use slower reduction implementation using temporaries
+        fn = sum_over_axis_strided_temps_dispatch_table[src_typeid][dst_typeid];
+        if (fn == nullptr) {
+            throw std::runtime_error("Datatypes are not supported");
+        }
+    }
+
+    std::vector<sycl::event> host_task_events{};
+    const auto &iter_src_dst_metadata_packing_triple_ =
+        dpctl::tensor::offset_utils::device_allocate_and_pack<py::ssize_t>(
+            exec_q, host_task_events, simplified_iteration_shape,
+            simplified_iteration_src_strides, simplified_iteration_dst_strides);
+    py::ssize_t *iter_shape_and_strides =
+        std::get<0>(iter_src_dst_metadata_packing_triple_);
+    if (iter_shape_and_strides == nullptr) {
+        throw std::runtime_error("Unable to allocate memory on device");
+    }
+    const auto &copy_iter_metadata_ev =
+        std::get<2>(iter_src_dst_metadata_packing_triple_);
+
+    const auto &reduction_metadata_packing_triple_ =
+        dpctl::tensor::offset_utils::device_allocate_and_pack<py::ssize_t>(
+            exec_q, host_task_events, simplified_reduction_shape,
+            simplified_reduction_src_strides);
+    py::ssize_t *reduction_shape_stride =
+        std::get<0>(reduction_metadata_packing_triple_);
+    if (reduction_shape_stride == nullptr) {
+        sycl::event::wait(host_task_events);
+        sycl::free(iter_shape_and_strides, exec_q);
+        throw std::runtime_error("Unable to allocate memory on device");
+    }
+    const auto &copy_reduction_metadata_ev =
+        std::get<2>(reduction_metadata_packing_triple_);
```

Comment:
Refactoring these to one device_allocate_and_pack call would improve stability on CPU